firmware/ubirch/m66/m66_http.c

```diff
@@ -63,6 +65,111 @@ int modem_http(http_method_t method, size_t *res_size, uint32_t timeout) {
   return status;
 }
 
+int modem_http_file_dl(const char *file_name, uint32_t timeout)
+{
+  timer_set_timeout(timeout * 1000);
+
+  int dl_size, content_len, dl_error_code = 0;
+
+  modem_send("AT+QHTTPDL=\"%s\"", file_name);
+
+  if (!modem_expect_OK(5 * 1000))
+  {
+    PRINTF("Failed to download file\r\n");
+    return 0;
+  }
+
+  if (modem_expect_scan("+QHTTPDL: %d,%d,%d", 10 * 5000, &dl_size, &content_len, &dl_error_code))
+  {
+    if (!dl_error_code)
+    {
+      PRINTF("Error downloading file: %d\r\n", dl_error_code);
+      return 0;
+    }
+
+    if (content_len == -1)
+    {
+      PRINTF("Content length unknown\r\n");
+      return 0;
+    }
+
+    if (!dl_size)
+    {
+      PRINTF("Downloaded ZERO bytes\r\n");
+      return 0;
+    }
+
+    //TODO: see if we need this print, either here or in main loop
+    PRINTF("dl_len = %d, content_len = %d, error_code = %d\r\n", dl_size, content_len, dl_error_code);
+  }
+
+  return dl_size;
+}
+
+int modem_http_file_open(const char *file_name, uint8_t rw_mode, uint32_t timeout)
+{
+  timer_set_timeout(timeout * 1000);
+
+  uint8_t file_handle;
+
+  modem_send("AT+QFOPEN=\"%s\",%d", file_name, rw_mode);
+
+  if (!modem_expect_scan("+QFOPEN: %u", uTimer_Remaining, &file_handle))
+  {
+    PRINTF("No File handle received\r\n");
+    return 0;
+  }
+
+  PRINTF("File handle number %u\r\n", file_handle);
+
+  return file_handle;
+}
+
+// TODO: add seek and position function, to get to the point where we need to start reading from
+size_t modem_http_file_read(char *read_buffer, int file_handle, size_t len, uint32_t timeout)
+{
+  timer_set_timeout(timeout * 1000);
+
+  char expect_connect[11] = {0};
+  snprintf(expect_connect, 11, "CONNECT %d", len);
+
+  modem_send("AT+QFREAD=%d,%d", file_handle, 100);
+
+  if (!modem_expect(expect_connect, uTimer_Remaining))
```

Comment:
shouldn't this be `if(!modem_expect_scan("CONNECT %d", &expected_len, ...)`
then check if expected_len == len